crates/iota-transaction-builder/src/builder/mod.rs

```diff
@@ -0,0 +1,955 @@
+// Copyright 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Builder for Programmable Transactions.
+
+use std::{
+    collections::{BTreeMap, HashMap},
+    marker::PhantomData,
+};
+
+use iota_crypto::IotaSigner;
+use iota_graphql_client::{Client, DryRunResult};
+use iota_types::{
+    Address, GasPayment, Identifier, ObjectId, ObjectReference, Owner, ProgrammableTransaction,
+    Transaction, TransactionEffects, TransactionExpiration, TypeTag,
+};
+use serde::Serialize;
+
+use crate::{
+    builder::{
+        named_commands::{NamedCommand, NamedCommands},
+        ptb_arguments::PTBArguments,
+    },
+    error::Error,
+    publish_type::PublishType,
+    types::{MoveType, MoveTypes},
+    unresolved::{
+        Argument, Command, Input, InputId, InputKind, MakeMoveVector, MergeCoins, MoveCall,
+        Publish, SplitCoins, TransferObjects, Upgrade,
+    },
+};
+
+mod named_commands;
+pub(crate) mod ptb_arguments;
+
+/// A transaction builder which can be used to construct [`Transaction`]s.
+#[derive(Debug, Clone)]
+#[repr(C)]
+pub struct TransactionBuilder<C = (), L = ()> {
+    data: TransactionBuildData,
+    client: C,
+    last_command: PhantomData<L>,
+}
+
+/// Transaction data used to build a [`Transaction`].
+#[derive(Debug, Clone)]
+pub struct TransactionBuildData {
+    /// The inputs to the transaction.
+    inputs: BTreeMap<InputId, Input>,
+    /// The list of commands in the transaction. A command is a single operation
+    /// in a programmable transaction.
+    commands: Vec<Command>,
+    /// The gas budget for the transaction.
+    gas_budget: Option<u64>,
+    /// The gas price for the transaction.
+    gas_price: Option<u64>,
+    /// The sender of the transaction.
+    sender: Address,
+    /// The sponsor of the transaction. If None, the sender is also the sponsor.
+    sponsor: Option<Address>,
+    /// The expiration of the transaction. The default value of this type is no
+    /// expiration.
+    expiration: TransactionExpiration,
+    named_commands: HashMap<String, Argument>,
+}
+
+impl TransactionBuildData {
+    fn set_input(&mut self, kind: InputKind, is_gas: bool) -> Argument {
```

Comment:
Any reason to not have this public?
I tried to use the clock as argument and did this
```Rust
    let clock = builder.input(iota_types::Input::Shared {
        object_id: "0x6".parse()?,
        initial_shared_version: 1,
        mutable: false,
    });
    builder
        .move_call(Address::TWO, "clock", "timestamp_ms")
        .arguments(clock)
        .name("timestamp_ms");
```
Or maybe even have a method like `.set_shared_input()` so one also doesn't have to care about the is_gas param?
Because now I have to resolve the shared object initial_shared_version myself first or did I miss something?

Reply:
You can just use the id `0x6`

Reply:
How?

Reply:
Ohh, you also need to wrap it like `Mut(clock)`. Sorry, missed that

Reply:
Doesn't need to be mutable for this call though `Error: Query error: [ErrorObject { code: InvalidParams, message: "Error checking transaction input objects: ImmutableParameterExpected { object_id: 0x0000000000000000000000000000000000000000000000000000000000000006 }", data: None }]`